azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,708 @@
+                                      _iot_device_module_list,
+                                      _iot_device_module_show,
+                                      _iot_device_module_twin_show,
+                                      _iot_device_module_twin_update,
+                                      _iot_device_set_parent, _iot_device_show,
+                                      _iot_device_twin_list,
+                                      _iot_device_twin_update,
+                                      _iot_edge_set_modules,
+                                      _iot_hub_configuration_create,
+                                      _iot_hub_configuration_delete,
+                                      _iot_hub_configuration_list)
+
+logger = get_logger(__name__)
+cli = EmbeddedCLI()
+
+
+class StateProvider(IoTHubProvider):
+    def __init__(
+        self,
+        cmd,
+        hub: Optional[str] = None,
+        rg: Optional[str] = None,
+        login: Optional[str] = None,
+        auth_type_dataplane: Optional[str] = None,
+        export: bool = False,
+    ):
+        try:
+            super(StateProvider, self).__init__(
+                cmd=cmd,
+                hub_name=hub,
+                rg=rg,
+                login=login,
+                auth_type_dataplane=auth_type_dataplane
+            )
+        except ResourceNotFoundError as e:
+            if export:
+                raise e
+            self.target = None
+            self.resolver = None
+        self.auth_type = auth_type_dataplane
+        self.login = login
+
+        if not self.hub_name:
+            self.hub_name = self.target["entity"].split('.')[0]
+        if not self.rg and not self.target:
+            raise RequiredArgumentMissingError(constants.MISSING_RG_ON_CREATE_ERROR)
+        if not self.rg and self.target:
+            self.rg = self.target.get("resourcegroup")
+
+    def _get_client(self):
+        return iot_hub_service_factory(self.cmd.cli_ctx)
+
+    def save_state(self, state_file: str, replace: bool = False, hub_aspects: List[str] = None):
+        '''Writes hub state to file'''
+        if (
+            os.path.exists(state_file)
+            and os.stat(state_file).st_size
+            and not replace
+            and not prompt_y_n(msg=constants.OVERWRITE_FILE_MSG.format(state_file), default="n")
+        ):
+            raise FileOperationError(constants.FILE_NOT_EMPTY_ERROR)
+
+        if not hub_aspects:
+            hub_aspects = HubAspects.list()
+        if HubAspects.Arm.value in hub_aspects and self.login:
+            raise MutuallyExclusiveArgumentError(constants.LOGIN_WITH_ARM_ERROR)
+
+        hub_state = self.process_hub_to_dict(self.target, hub_aspects)
+
+        try:
+            with open(state_file, 'w', encoding='utf-8') as f:
+                json.dump(hub_state, f, indent=4, sort_keys=True)
+
+            logger.info(constants.SAVE_STATE_MSG.format(self.hub_name, state_file))
+
+        except FileNotFoundError:
```

Comment:
Also look for other opportunities of re-use around fileops

Reply:
other possible uses for file ops are prob certificate uploads.

Reply:
ok looking at the file ops - I have some issues:
1. the current save state checks if the file is present and the user is willing to overwrite before the full state is fetched. This makes sense to me since it is a quick operation (rather than having the cli download 1k devices just to realize that the file is already in use)
2. the write_content_to_file writes things in bytes. Meanwhile I want to use json.dump so I can dump the json with pretty print that is easy for users to look at and compare. So I will have to write another function or make write_content_to_file more complex.

Reply:
1. Ok don't use it if it doesn't fit.
2. This is not really any issue, the function takes content of str or bytes...you can format with `json.dumps(...,indent=4)` and pass in the formatted string.